File: src/main/java/org/gabi/projet_gl/model/Project.java
package org.gabi.projet_gl.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

@Entity
public class Project {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long project_id;
  //code pour l'association many to many
  @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
  @JoinTable(
      name = "project_participants",
      joinColumns = @JoinColumn(name = "project_id"),
      inverseJoinColumns = @JoinColumn(name = "user_id")
  )
  @JsonBackReference
  private Set<AppUser> participants;
  @JsonIgnore
  @OneToMany(mappedBy = "project")
  private Set<Task> tasks;
  private String name;
  private String description;
  private boolean done;
  public LocalDate getStartDate() {
    return startDate;
  }
  public void setStartDate(LocalDate startDate) {
    this.startDate = startDate;
  }
  public LocalDate getDeadline() {
    return deadline;
  }
  public void setDeadline(LocalDate deadline) {
    this.deadline = deadline;
  }
  private LocalDate startDate;
  private LocalDate deadline;
  public String getName() {
    return name;
  }
  public void setName(String name) {
    this.name = name;
  }
  public String getDescription() {
    return description;
  }
  public void setDescription(String description) {
    this.description = description;
  }
  public boolean isDone() {
    return done;
  }
  public void setDone(boolean done) {
    this.done = done;
  }
  public Set<Task> getTasks() {
    return tasks;
  }
  public void setTasks(Set<Task> tasks) {
    this.tasks = tasks;
  }



  public Project(Long project_id,
                 String name,
                 String description,
                 boolean done,
                 LocalDate startDate,
                 LocalDate deadline)
  {
    this.project_id = project_id;
    this.name = name;
    this.description = description;
    this.done = done;
    this.startDate = startDate;
    this.deadline = deadline;
    this.participants = new HashSet<>();
    this.tasks = new HashSet<>();
  }

  public Project() {}
  public void setId(Long project_id) {
    this.project_id = project_id;
  }

  public Long getProject_id() {
    return this.project_id;
  }
  public Set<AppUser> getParticipants() {
    return participants;
  }

}

File: frontend/src/app/project-class/project/project.ts
import {User} from '../../user-class/user/user';
import {Task} from '../../task-class/task/task';


export class Project{
  private _project_id : string = "";
  private _name : string = "New Project";
  private _done : boolean = false;
  private _startDate : Date = new Date("0000-00-00");
  private _deadline : Date = new Date("0000-00-00");
  private _description : string = "";
  private _participants : User[] = [];
  private _tasks : Task[] = [];
  private _color : string = "#" + Math.floor(Math.random()*16777215).toString(16);

  constructor(){

  }

  get project_id() : string{
    return this._project_id;
  }
  set project_id(project_id : string){
    this._project_id = project_id;
  }

  get name() : string{
    return this._name;
  }
  set name(name : string){
    this._name = name;
  }

  get done() : boolean{
    return this._done;
  }
  set done(done : boolean){
    this._done = done;
  }

  get startDate() : Date{
    return this._startDate;
  }
  set startDate(startDate : Date){
    this._startDate = startDate;
  }

  get deadline() : Date{
    return this._deadline;
  }
  set deadline(deadLine : Date){
    this._deadline = deadLine;
  }

  get description(): string {
    return this._description;
  }

  set description(value: string) {
    this._description = value;
  }

  get participants() : User[]{
    return this._participants;
  }
  set participants(participants : User[]){
    this._participants = participants;
  }

  get tasks() : Task[]{
    return this._tasks;
  }
  set tasks(tasks : Task[]){
    this._tasks = tasks;
  }

  get color(): string {
    return this._color;
  }

  set color(value: string) {
    this._color = value;
  }
}
